src/test/java/com/ita/if103java/ims/service/WarehouseServiceImplTest.java

```diff
@@ -0,0 +1,454 @@
+                1L, true));
+        when(warehouseDao.findByTopWarehouseID(1L, 1L)).thenReturn(List.of(warehouse));
+        when(warehouseDtoMapper.toDtoList(warehouseList)).thenReturn(List.of(warehouseDto));
+    }
+
+
+    @Test
+    public void softDeleteTest_throwExceptionWhenSubWarehouseExists() {
+        Warehouse targetWarehouse = new Warehouse();
+
+        when(warehouseDao.findById(1L, 1L)).thenReturn(targetWarehouse);
+        when(warehouseDao.findChildrenById(1L, 1L)).thenReturn(List.of(new Warehouse()));
+
+        WarehouseDeleteException warehouseDeleteException = assertThrows(WarehouseDeleteException.class, () -> {
+            warehouseService.softDelete(1L, userDetails);
+        });
+        assertEquals("Warehouse has sub warehouses! Firstly you should delete them!", warehouseDeleteException.getMessage());
+    }
+
+    @Test
+    public void softDeleteTest_throwExceptionWhenSavedItemExists() {
+        Warehouse targetWarehouse = new Warehouse();
+
+        when(warehouseDao.findById(1L, 1L)).thenReturn(targetWarehouse);
+        when(warehouseDao.findChildrenById(1L, 1L)).thenReturn(Collections.emptyList());
+        when(savedItemDao.findSavedItemByWarehouseId(1L)).thenReturn(List.of(new SavedItem()));
+
+        WarehouseDeleteException warehouseDeleteException = assertThrows(WarehouseDeleteException.class, () -> {
+            warehouseService.softDelete(1L, userDetails);
+        });
+        assertEquals("Warehouse is not empty! Firstly you should remove or transfer all items from that warehouse to another", warehouseDeleteException.getMessage());
+    }
+
+    @Test
+    public void softDeleteTest_warehouseIsNotDeleted() {
+        Warehouse targetWarehouse = new Warehouse();
+
+        when(warehouseDao.findById(1L, 1L)).thenReturn(targetWarehouse);
+        when(warehouseDao.findChildrenById(1L, 1L)).thenReturn(Collections.emptyList());
+        when(savedItemDao.findSavedItemByWarehouseId(1L)).thenReturn(Collections.emptyList());
+        when(warehouseDao.softDelete(1L)).thenReturn(false);
+
+        boolean result = warehouseService.softDelete(1L, userDetails);
+
+        assertFalse(result);
+
+        verify(eventService, never()).create(any(Event.class));
+    }
+
+    @Test
+    public void softDeleteTest_warehouseIsDeleted() {
+        Warehouse targetWarehouse = new Warehouse();
+
+        when(warehouseDao.findById(1L, 1L)).thenReturn(targetWarehouse);
+        when(warehouseDao.findChildrenById(1L, 1L)).thenReturn(Collections.emptyList());
+        when(savedItemDao.findSavedItemByWarehouseId(1L)).thenReturn(Collections.emptyList());
+        when(warehouseDao.softDelete(1L)).thenReturn(true);
+
+        boolean result = warehouseService.softDelete(1L, userDetails);
+
+        assertTrue(result);
+
+        verify(eventService).create(any(Event.class));
+    }
+
+    @Test
+    public void findTotalCapacityTest() {
+        when(warehouseDao.findTotalCapacity(1L, 2L)).thenReturn(5);
+
+    }
```

Comment:
this test test nothing, there is no invocation of the warehouseService
Fix or remove